src/Compilers/CSharp/Portable/FlowAnalysis/VariableUsePass.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// VariableUsePass marks all referenced variables in the visited node as used (as a read).
+    /// This pass is used for things like attribute arguments and default parameter expressions,
+    /// which are otherwise not analyzed for use references.
+    /// </summary>
+    /// <remarks>
+    /// An instance of this class is cached on CSharpCompilation, as it is stateless (other than the SoruceAssembly).
+    /// Be careful adding new state (either avoid state that cannot be shared across the compilation, or do not cache an instance on the compilation).
+    /// </remarks>
+    internal class VariableUsePass : BoundTreeWalkerWithStackGuard
+    {
+        /// <summary>
+        /// The current source assembly.
+        /// </summary>
+        private readonly SourceAssemblySymbol _sourceAssembly;
+
+        /// <summary>
+        /// Variables that were used anywhere, in the sense required to suppress warnings about
+        /// unused variables.
+        /// </summary>
+        private readonly HashSet<LocalSymbol> _usedVariablesOpt;
+
+        /// <summary>
+        /// Variables that were used anywhere, in the sense required to suppress warnings about
+        /// unused variables.
+        /// </summary>
+        private readonly HashSet<LocalFunctionSymbol> _usedLocalFunctionsOpt;
```

Comment:
This is a bit confusing - the class is used both for field references and local variable references. The one that's cached on the CSharpCompilation has these two fields be null (and not used). The one that's used in the DataFlowPass is created per-dataflowpass and has these be non-null. Would you suggest documenting this better, or creating a subclass with these two fields? (and using the base for the global one, and derived for the DataFlowPass)